File: MarxUp/MarxUp/Annotator.m
//
//  Annotator.m
//  MarxUp
//
//  Created by A-Team User on 1.10.18.
//  Copyright © 2018 Ognyanka Boneva. All rights reserved.
//

#import "Annotator.h"
#import "ChangesHistory.h"

#import "Utilities.h"
#import "Constants.h"

#define ROUNDED_RECT_CORNER_RADIUS 0.5

@interface Annotator ()

@property (strong, nonatomic) ChangesHistory *annotationsHistory;
@property (assign) ContentType annotatedContent;
@property (assign) ToolboxItemType choosenItem;

@property (strong, nonatomic) UIColor *color;
@property (assign) CGFloat lineWidth;
@property (assign) PDFLineStyle endLineStyle;
@property (assign) ShapeType shape;
@property (strong, nonatomic) UIBezierPath *path;

@property (strong, nonatomic) PDFAnnotation *annotation;
@property (assign) BOOL isAnnotationAdded;
@property (strong, nonatomic) PDFView *annotatedPDF;

@property (strong, nonatomic) UIImage *originalImage;
@property (strong, nonatomic) UIImage *changedImage;
@property (strong, nonatomic) UIImageView *annotatedImage;

@property (assign) CGPoint lastPoint;
@property (assign) CGPoint startPoint;

@end

@implementation Annotator

+ (instancetype)newForAnnotatingImage:(UIImageView *)imageView {
    Annotator *new = [Annotator new];
    new.annotatedImage = imageView;
    new.originalImage = imageView.image;
    new.changedImage = imageView.image.copy;
    [new setDefultValuesToProperties];
    new.annotationsHistory = [ChangesHistory newWithChangeType:UIImage.class];
    new.annotatedContent = ContentTypeImage;
    
    return new;
}

+ (instancetype)newForAnnotatingPDF:(PDFView *)pdfView {
    Annotator *new = [Annotator new];
    new.annotatedPDF = pdfView;
    [new setDefultValuesToProperties];
    new.annotationsHistory = [ChangesHistory newWithChangeType:PDFAnnotation.class];
    new.annotatedContent = ContentTypePDF;
    
    return new;
}

- (void)beginAnnotatingAtPoint:(CGPoint)point {
    self.lastPoint = point;
    self.startPoint = point;
    if (self.annotatedContent == ContentTypeImage) {
        UIGraphicsBeginImageContext(self.annotatedImage.image.size);
        [self.changedImage drawAtPoint:CGPointZero];
    }
    if (self.choosenItem != ToolboxItemTypeArrow) {
        [self beginDrawingWithBezierPathAtPoint:point];
    }
    
    self.isAnnotationAdded = false;
}

- (void)endAnnotatingAtPoint:(CGPoint)point {
    [self continueAnnotatingAtPoint:point];
    if (self.annotatedContent == ContentTypeImage) {
        self.changedImage = UIGraphicsGetImageFromCurrentImageContext();
        UIGraphicsEndImageContext();
    }
    [self.annotationsHistory addChange:self.annotation];
}

- (void)continueAnnotatingAtPoint:(CGPoint)point {
    [self updateBezierPathWithPoint:point];
    [self removePreviousVersionOfAnnotationWithPoint:point];
    
    if (self.annotatedContent == ContentTypeImage) {
        [self strokePathOverImage];
    }
    else {
        [self addAnnotationWithCurrentBezierPath];
        [self strokePath];
    }
    
    self.lastPoint = point;
}

- (void)strokePathOverImage {
    [self strokePath];
    self.annotatedImage.image = UIGraphicsGetImageFromCurrentImageContext();
}

- (void)setDefultValuesToProperties {
    self.color = UIColor.blackColor;
    self.lineWidth = 3;
    self.choosenItem = ToolboxItemTypePen;
}

- (void)beginDrawingWithBezierPathAtPoint:(CGPoint)point {
    self.path = [UIBezierPath new];
    self.path.lineWidth = self.lineWidth;
    [self.path moveToPoint:point];
}

- (void)updatePropertie:(NSInteger)property fromType:(ToolboxItemType)type {
    if (type == ToolboxItemTypeWidth) {
        self.lineWidth = property;
    }
    else if (type == ToolboxItemTypeArrow && property == ArrowEndLineTypeOpen) {
        self.endLineStyle = kPDFLineStyleOpenArrow;
    }
    else if (type == ToolboxItemTypeArrow && property == ArrowEndLineTypeClosed) {
        self.endLineStyle = kPDFLineStyleClosedArrow;
    }
    else if (type == ToolboxItemTypeShape) {
        self.shape = property;
    }
}

- (void)updatePropertiesForAnnotation {
    self.annotation.color = self.color;
    self.annotation.border.lineWidth = self.lineWidth;
}

- (void)updateBezierPathWithPoint:(CGPoint)point {
    if (self.choosenItem == ToolboxItemTypePen) {
        [self.path addLineToPoint:point];
    }
    else if (self.choosenItem == ToolboxItemTypeShape) {
        [self addShapeWithBezierPathAtPoint:point];
    }
    else if (self.choosenItem == ToolboxItemTypeArrow) {
        [self addArrowWithBezierPathAtPoint:point];
    }
}

-(void)removePreviousVersionOfAnnotationWithPoint:(CGPoint)point {
    if (self.annotatedContent == ContentTypePDF) {
        if (self.isAnnotationAdded) {
            [self.annotatedPDF.currentPage removeAnnotation:self.annotation];
        }
        else {
            self.lastPoint = point;
            self.isAnnotationAdded = true;
        }
    }
    else if (self.annotatedContent == ContentTypeImage) {
        [self.changedImage drawAtPoint:CGPointZero];
    }
}

- (void)addAnnotationWithCurrentBezierPath {
    CGRect annotationBounds = [self.annotatedPDF.currentPage boundsForBox:kPDFDisplayBoxMediaBox];
    self.annotation = [[PDFAnnotation alloc] initWithBounds:annotationBounds forType:PDFAnnotationSubtypeInk withProperties:nil];
    [self updatePropertiesForAnnotation];
    [self.annotation addBezierPath:self.path];
    [self.annotatedPDF.currentPage addAnnotation:self.annotation];
}
    
- (void)addShapeWithBezierPathAtPoint:(CGPoint)point {
    CGRect shapeRect = [Utilities rectBetweenPoint:point andOtherPoint:self.startPoint];
    
    if (self.shape == ShapeTypeRegularRectangle) {
        self.path = [UIBezierPath bezierPathWithRect:shapeRect];
    }
    else if (self.shape == ShapeTypeCircle) {
        self.path = [UIBezierPath bezierPathWithOvalInRect:shapeRect];
    }
    else if (self.shape == ShapeTypeRoundedRectangle) {
        self.path = [UIBezierPath bezierPathWithRoundedRect:shapeRect cornerRadius:ROUNDED_RECT_CORNER_RADIUS];
    }
    self.path.lineWidth = self.lineWidth;
}

- (void)addArrowWithBezierPathAtPoint:(CGPoint)point {
    self.path = [UIBezierPath bezierPath];
    [self.path moveToPoint:self.startPoint];
    [self.path addLineToPoint:point];
    
    CGPoint pointForVerticalLine = CGPointMake(point.x, self.startPoint.y + (point.y - self.startPoint.y) * 0.3);
    CGPoint pointforHorizontalLine = CGPointMake(self.startPoint.x + (point.x - self.startPoint.x) * 0.3, point.y);
    if (point.y < self.startPoint.y) {
            pointForVerticalLine = CGPointMake(point.x, point.y + (self.startPoint.y - point.y) * 0.3);
    }
    if(point.x < self.startPoint.x) {
            pointforHorizontalLine = CGPointMake(point.x + (self.startPoint.x - point.x) * 0.3, point.y);
    }
    
    [self.path moveToPoint:point];
    [self.path addLineToPoint:pointForVerticalLine];
    
    [self.path moveToPoint:point];
    [self.path addLineToPoint:pointforHorizontalLine];
    
    if (self.endLineStyle == kPDFLineStyleClosedArrow) {
        UIGraphicsBeginImageContext(self.annotatedPDF.frame.size);
        
        [self.path moveToPoint:pointforHorizontalLine];
        [self.path addLineToPoint:pointForVerticalLine];
        
        [self.color setFill];
        [self.path closePath];
        [self.path fill];
        
        UIGraphicsEndImageContext();
    }
    
    self.path.lineWidth = self.lineWidth;
}

- (void)addAnnotationFromSubtype:(PDFAnnotationSubtype)subtype withMarkUpType:(PDFMarkupType)markUp {
    for (PDFSelection *lineSelection in self.annotatedPDF.currentSelection.selectionsByLine) {
        self.annotation = [[PDFAnnotation alloc] initWithBounds:[lineSelection boundsForPage:self.annotatedPDF.currentPage] forType:subtype withProperties:nil];
        [self.annotation setMarkupType:markUp];
        [self updatePropertiesForAnnotation];
        [self.annotatedPDF.currentPage addAnnotation:self.annotation];
        [self.annotationsHistory addChange:self.annotation];
    }
}

- (void)strokePath {
    [self.color setStroke];
    self.path.lineWidth = self.lineWidth;
    [self.path stroke];
}

- (UIImage *)getOriginalImage {
    return self.originalImage;
}

- (UIImage *)getCurrentImage {
    return UIGraphicsGetImageFromCurrentImageContext();
}

- (void)didChooseTextAnnotationFromType:(ToolboxItemType)type {
    if (type == ToolboxItemTypeTextHighlight) {
        [self addAnnotationFromSubtype:PDFAnnotationSubtypeHighlight withMarkUpType:kPDFMarkupTypeHighlight];
    }
    else if (type == ToolboxItemTypeTextStrikeThrough) {
        [self addAnnotationFromSubtype:PDFAnnotationSubtypeStrikeOut withMarkUpType:kPDFMarkupTypeStrikeOut];
    }
    else if (type == ToolboxItemTypeTextUnderline) {
        [self addAnnotationFromSubtype:PDFAnnotationSubtypeUnderline withMarkUpType:kPDFMarkupTypeUnderline];
    }
}

- (void)didChoosePen {
    self.choosenItem = ToolboxItemTypePen;
}

- (void)didChooseOption:(NSInteger)option forType:(ToolboxItemType)itemType {
    if (itemType == ToolboxItemTypeShape || itemType == ToolboxItemTypeArrow) {
        self.choosenItem = itemType;
    }
    [self updatePropertie:option fromType:itemType];
}

- (void)didChooseColor:(UIColor *)color {
    self.color = color;
}

- (void)didChooseLineWidth:(CGFloat)width {
    self.lineWidth = width;
}

- (void)didSelectRedo {
    if ([self.annotationsHistory redoChange]) {
        [self.annotatedPDF.currentPage addAnnotation:[self.annotationsHistory lastChange]];
    }
}

- (void)didSelectUndo {
    if ([self.annotationsHistory undoChange]) {
        [self.annotatedPDF.currentPage removeAnnotation:[self.annotationsHistory lastInvisibleChange]];
    }
}

- (void)clearChanges {
    [self.annotationsHistory cleanAllChanges];
}

- (void)reset {
    if (self.annotatedContent == ContentTypePDF) {
        for (PDFAnnotation *annotation in [self.annotationsHistory getAllChanges]) {
            [self.annotatedPDF.currentPage removeAnnotation:annotation];
        }
    }
    else {
        self.annotatedImage.image = self.originalImage;
    }
    [self clearChanges];
}

@end

File: MarxUpUp/MarxUpUp/Annotations/Annotator.swift
//
//  Annotator.swift
//  MarxUpUp
//
//  Created by Ognyanka Boneva on 9.11.18.
//  Copyright © 2018 Ognyanka Boneva. All rights reserved.
//

import UIKit
import PDFKit

class Annotator: NSObject {

    private var content: ContentType!
    private var choosenItem: ToolboxItemType = ToolboxItemType.pen

    private var color: UIColor = UIColor.black
    private var lineWidth: CGFloat = 2.0
    private var endLineStyle: ArrowEndLineType = ArrowEndLineType.open
    private var shape: ShapeType = ShapeType.circle
    private var path: UIBezierPath = UIBezierPath()

    private var annotation: PDFAnnotation?
    private var isAnnotationAdded: Bool = false
    var annotatedPDFView: PDFView?

    var annotatedImage: UIImage?
    private var annotatedImageView: UIImageView?
    private var currentImage: UIImage? {
        return UIGraphicsGetImageFromCurrentImageContext()
    }
    private var shouldAnnotate: Bool {
        return (content == ContentType.image && annotatedImage != nil) ||
                (content == ContentType.pdf && annotatedPDFView != nil)
    }

    private var startPoint: CGPoint!
    private var lastPoint: CGPoint!

    private let changesManager = ChangesManager()

    init(forAnnotating imageView: UIImageView) {
        content = ContentType.image

        guard imageView.image != nil else {
            print("Error: There is no image to annotate.")
            return
        }
        annotatedImageView = imageView
        annotatedImage = imageView.image
    }

    init(forAnnotating PDFView: PDFView) {
        content = ContentType.pdf

        guard PDFView.document != nil else {
            print("Error: There is no document to annotate.")
            return
        }
        annotatedPDFView = PDFView
    }

    // MARK: - BezierPaths based annotating
    private func beginDrawingWithBezierPath(atPoint point: CGPoint) {
        path = UIBezierPath()
        path.lineWidth = lineWidth
        path.move(to: point)
    }

    private func updateBezierPath(withPoint point: CGPoint) {
        if choosenItem == ToolboxItemType.pen {
            path.addLine(to: point)
        } else if choosenItem == ToolboxItemType.shape {
            addShapeWithBezierPath(atPoint: point)
        } else if choosenItem == ToolboxItemType.arrow {
            addArrowWithBezierPath(atPoint: point)
        }
    }

    private func addArrowWithBezierPath(atPoint point: CGPoint) {
        let angle = Utilities.angleBetweenPoint(point, andOtherPoint: startPoint)
        path = ArrowBezierPath.endLine(atPoint: point, fromType: endLineStyle)
        Utilities.rotateBezierPath(path, aroundPoint: point, withAngle: angle)

        path.move(to: startPoint)
        path.addLine(to: CGPoint(x: point.x, y: point.y))

        path.lineWidth = lineWidth
    }

    private func addShapeWithBezierPath(atPoint point: CGPoint) {
        let shapeRect = Utilities.rectBetween(point, startPoint)

        if shape == ShapeType.regularRectangle {
            path = UIBezierPath(rect: shapeRect)
        } else if shape == ShapeType.roundedRectangle {
            path = UIBezierPath(roundedRect: shapeRect, cornerRadius: 2.0)
        } else if shape == ShapeType.circle {
            path = UIBezierPath(ovalIn: shapeRect)
        }

        path.lineWidth = lineWidth
    }

    private func addAnnotationWithCurrentBezierPath() {
        guard let annotationBounds = annotatedPDFView?.currentPage?.bounds(for: PDFDisplayBox.mediaBox) else {
            return
        }
        annotation = PDFAnnotation(bounds: annotationBounds, forType: PDFAnnotationSubtype.ink, withProperties: nil)
        updatePropertiesForAnnotation()
        annotation?.add(path)
        if annotation != nil {
            annotatedPDFView?.currentPage?.addAnnotation(annotation!)
        }
    }

    // MARK: - Updating properties
    private func updateProperty(_ property: Int, fromType itemType: ToolboxItemType) {
        if itemType == ToolboxItemType.width {
            lineWidth = CGFloat(property)
        } else if itemType == ToolboxItemType.arrow && property == ArrowEndLineType.open.rawValue {
            endLineStyle = ArrowEndLineType.open
        } else if itemType == ToolboxItemType.arrow && property == ArrowEndLineType.closed.rawValue {
            endLineStyle = ArrowEndLineType.closed
        } else if itemType == ToolboxItemType.shape {
            guard let newShape = ShapeType(rawValue: property) else {
                return
            }
            shape = newShape
        }
    }

    private func updatePropertiesForAnnotation() {
        annotation?.color = color
        annotation?.border?.lineWidth = lineWidth
    }

    // MARK: - Visualize annotation
    private func strokePath() {
        color.setStroke()
        color.setFill()
        path.lineWidth = lineWidth
        if endLineStyle == .closed { path.fill() }

        path.stroke()

        if content == ContentType.image {
            annotatedImageView?.image = currentImage
        }
    }

    private func addAnnotation(_ sybtype: PDFAnnotationSubtype, markUpType: PDFMarkupType) {
        var annotations = [PDFAnnotation]()
        annotatedPDFView?.currentSelection?.selectionsByLine().forEach({ selection in
            if annotatedPDFView != nil, annotatedPDFView!.currentPage != nil {
                annotation = PDFAnnotation(bounds: selection.bounds(for: annotatedPDFView!.currentPage!),
                                           forType: sybtype, withProperties: nil)
                annotation?.markupType = markUpType
                updatePropertiesForAnnotation()
                annotatedPDFView?.currentPage?.addAnnotation(annotation!)
                annotations.append(annotation!)
            }
        })

        guard let page = annotatedPDFView?.currentPage else {
            return
        }
        changesManager.addTextAnnotation(annotations, forPage: page)
    }

    // MARK: - Others
    private func removePreviousVersionOfAnnotation(withPoint point: CGPoint) {
        if content == ContentType.pdf, isAnnotationAdded, annotation != nil {
            annotatedPDFView?.currentPage?.removeAnnotation(annotation!)
        } else if content == ContentType.pdf {
            lastPoint = point
            isAnnotationAdded = true
        } else {
            annotatedImage?.draw(at: CGPoint.zero)
        }
    }
}

// MARK: - Annotating Methods
extension Annotator: Annotating {

    var isThereUnsavedWork: Bool {
        return changesManager.unsavedWork
    }

    func beginAnnotating(atPoint point: CGPoint) {
        if !shouldAnnotate {
            return
        }

        lastPoint = point
        startPoint = point

        if content == ContentType.image {
            guard let size = annotatedImageView?.image?.size else {
                return
            }
            UIGraphicsBeginImageContext(size)
            annotatedImage?.draw(at: CGPoint.zero)
        }
        if choosenItem != ToolboxItemType.arrow {
            beginDrawingWithBezierPath(atPoint: point)
        }

        isAnnotationAdded = false
    }

    func continueAnnotating(atPoint point: CGPoint ) {
        if !shouldAnnotate {
            return
        }

        updateBezierPath(withPoint: point)
        removePreviousVersionOfAnnotation(withPoint: point)

        if content == ContentType.pdf {
            addAnnotationWithCurrentBezierPath()
        }

        strokePath()
        lastPoint = point
    }

    func endAnnotating(atPoint point: CGPoint) {
        if !shouldAnnotate {
            return
        }

        continueAnnotating(atPoint: point)

        if content == ContentType.image {
            let fill = (endLineStyle == .closed)
            changesManager.addInkImageAnnotation(withLines: path, forImage: annotatedImage!, andFill: fill)
            annotatedImage = currentImage
            UIGraphicsEndImageContext()
        } else {
            guard let page = annotatedPDFView?.currentPage, annotation != nil else {
                return
            }
            changesManager.addInkPDFAnnotation(annotation!, forPage: page)
        }
    }

    func reset() {
        if UIGraphicsGetCurrentContext() == nil, let size = annotatedImage?.size {
            UIGraphicsBeginImageContext(size)
        }
        annotatedImage?.draw(at: CGPoint.zero)
        changesManager.reset()
        annotatedImageView?.image = currentImage
        annotatedImage = currentImage
    }

    func save() {
        changesManager.clear()
        if UIGraphicsGetCurrentContext() == nil, let size = annotatedImage?.size {
            UIGraphicsBeginImageContext(size)
        }
        annotatedImage?.draw(at: CGPoint.zero)
        annotatedImageView?.image = currentImage
        annotatedImage = currentImage
    }
}

// MARK: - ToolboxItemDelegate Methods
extension Annotator: ToolboxItemDelegate {
    func didChoose(textAnnotationFromType type: ToolboxItemType) {
        if type == ToolboxItemType.textHighlight {
            addAnnotation(PDFAnnotationSubtype.highlight, markUpType: PDFMarkupType.highlight)
        } else if type == ToolboxItemType.textUnderline {
            addAnnotation(PDFAnnotationSubtype.underline, markUpType: PDFMarkupType.underline)
        } else if type == ToolboxItemType.textStrikeThrough {
            addAnnotation(PDFAnnotationSubtype.strikeOut, markUpType: PDFMarkupType.strikeOut)
        }
    }

    func didChoosePen() {
        choosenItem = ToolboxItemType.pen
    }

    func didChoose(_ option: Int, forToolboxItem type: ToolboxItemType) {
        if type == ToolboxItemType.shape || type == ToolboxItemType.arrow {
            choosenItem = type
        }
        updateProperty(option, fromType: type)
    }

    func didChoose(lineWidth width: Float) {
        lineWidth = CGFloat(width)
    }

    func didChooseColor(_ color: UIColor) {
        self.color = color
    }
}

// MARK: - EditedContentStateDelegate Methods
extension Annotator: EditedContentStateDelegate {
    func didSelectRedo() {
        if UIGraphicsGetCurrentContext() == nil, let size = annotatedImage?.size {
            UIGraphicsBeginImageContext(size)
        }
        color.setStroke()
        color.setFill()
        changesManager.redo {
            annotatedImageView?.image = currentImage
            annotatedImage = currentImage
        }
    }

    func didSelectUndo() {
        if UIGraphicsGetCurrentContext() == nil, let size = annotatedImage?.size {
            UIGraphicsBeginImageContext(size)
        }
        changesManager.undo {
            annotatedImageView?.image = currentImage
            annotatedImage = currentImage
        }
    }
}
